pkg/controller/operators/olm/operatorgroup.go

```diff
@@ -981,61 +994,129 @@ func (a *Operator) updateNamespaceList(op *operatorsv1.OperatorGroup) ([]string,
-			MatchLabels: map[string]string{
-				aggregationLabel: "true",
-			},
-		})
-	}
-	if len(selectors) > 0 {
-		clusterRole.AggregationRule = &rbacv1.AggregationRule{
-			ClusterRoleSelectors: selectors,
-		}
+
+	if err := ownerutil.AddOwnerLabels(clusterRole, op); err != nil {
+		return err
 	}
-	err := ownerutil.AddOwnerLabels(clusterRole, op)
+
+	// list current cluster roles owned by this operator group
+	clusterRoles, err := a.lister.RbacV1().ClusterRoleLister().List(ownerutil.OperatorGroupOwnerSelector(op))
 	if err != nil {
 		return err
 	}
 
-	existingRole, err := a.lister.RbacV1().ClusterRoleLister().Get(clusterRole.Name)
-	if err != nil && !apierrors.IsNotFound(err) {
-		return err
+	// find the cluster role that matches the template and is of the correct type
+	var existingClusterRoles []*rbacv1.ClusterRole
+	re, ok := ogClusterRoleNameRegExMap[roleType]
+	if !ok {
+		return fmt.Errorf("no regex found for role type: %s", roleType)
 	}
-	if apierrors.IsNotFound(err) {
-		existingRole, err = a.opClient.KubernetesInterface().RbacV1().ClusterRoles().Create(context.TODO(), clusterRole, metav1.CreateOptions{})
-		if err == nil {
-			return nil
+	for _, cr := range clusterRoles {
+		if re.FindStringSubmatch(cr.GetName()) != nil {
+			existingClusterRoles = append(existingClusterRoles, cr)
 		}
-		if !apierrors.IsAlreadyExists(err) {
+	}
+
+	switch len(existingClusterRoles) {
+	// if no cluster role exists, create one
+	case 0:
+		a.logger.Infof("Creating cluster role: %s owned by operator group: %s/%s", clusterRole.GetGenerateName(), op.GetNamespace(), op.GetName())
+		if _, err = a.opClient.KubernetesInterface().RbacV1().ClusterRoles().Create(context.TODO(), clusterRole, metav1.CreateOptions{}); err != nil {
+			// name collision, try again with a new name in the next reconcile
 			a.logger.WithError(err).Errorf("Create cluster role failed: %v", clusterRole)
+		}
+		return err
+	// if an existing cluster role resource is found, update it if necessary
+	case 1:
+		existingClusterRole := existingClusterRoles[0].DeepCopy()
+		// the cluster role will need to be updated if the aggregation rules have changed - otherwise, nothing to do
+		// the resource is guaranteed to have the correct ownership labels because we reached this part of the code
+		if !reflect.DeepEqual(existingClusterRole.AggregationRule, aggregationRule) {
+			if _, err := a.opClient.UpdateClusterRole(existingClusterRole); err != nil {
+				a.logger.WithError(err).Errorf("Update existing cluster role failed: %v", clusterRole)
+			}
 			return err
 		}
-	}
+	// the inherent race condition created by listing to check if a resource exists and then creating it
+	// and the fact that we are using generateName means that it is possible for multiple cluster roles of the same
+	// role type to be created for the same operator group. In this case, we'll disambiguate by keeping the oldest
+	// cluster role (by creation timestamp - tie-breaking by name) and deleting the others
```

Comment:
These are the complications we were hoping to avoid, and also the names for the 3 operator group won't have the same identifier, making it harder. Ugh... I'm leaning toward something deterministic now... :sorry:

Reply:
All good. Reached the same conclusion

Reply:
Less to test too

Reply:
I assume this controller has leader election set up. If that's the case, there will only ever be a singleton holding the write lease for this controller's process, and there's nobody to race with on the creation of this object.

Reply:
what's not clear to me is whether the controller's process drains the queue serially or not

Reply:
It does not, but queue items get deduplicated